librz/analysis/dwarf_process.c

```diff
@@ -316,51 +322,66 @@ static st32 parse_type(Context *ctx, const ut64 offset, RzStrBuf *strbuf, ut64 *
 	case DW_TAG_array_type:
 		type_idx = find_attr_idx(die, DW_AT_type);
 		if (type_idx != -1) {
-			parse_type(ctx, die->attr_values[type_idx].reference, strbuf, size);
+			parse_type(ctx, die->attr_values[type_idx].reference, strbuf, size, visited);
 		}
 		parse_array_type(ctx, die - ctx->all_dies, strbuf);
 		break;
 	case DW_TAG_const_type:
 		type_idx = find_attr_idx(die, DW_AT_type);
 		if (type_idx != -1) {
-			parse_type(ctx, die->attr_values[type_idx].reference, strbuf, size);
+			parse_type(ctx, die->attr_values[type_idx].reference, strbuf, size, visited);
 		}
 		rz_strbuf_append(strbuf, " const");
 		break;
 	case DW_TAG_volatile_type:
 		type_idx = find_attr_idx(die, DW_AT_type);
 		if (type_idx != -1) {
-			parse_type(ctx, die->attr_values[type_idx].reference, strbuf, size);
+			parse_type(ctx, die->attr_values[type_idx].reference, strbuf, size, visited);
 		}
 		rz_strbuf_append(strbuf, " volatile");
 		break;
 	case DW_TAG_restrict_type:
 		type_idx = find_attr_idx(die, DW_AT_type);
 		if (type_idx != -1) {
-			parse_type(ctx, die->attr_values[type_idx].reference, strbuf, size);
+			parse_type(ctx, die->attr_values[type_idx].reference, strbuf, size, visited);
 		}
 		rz_strbuf_append(strbuf, " restrict");
 		break;
 	case DW_TAG_rvalue_reference_type:
 		type_idx = find_attr_idx(die, DW_AT_type);
 		if (type_idx != -1) {
-			parse_type(ctx, die->attr_values[type_idx].reference, strbuf, size);
+			parse_type(ctx, die->attr_values[type_idx].reference, strbuf, size, visited);
 		}
 		rz_strbuf_append(strbuf, " &&");
 		break;
 	case DW_TAG_reference_type:
 		type_idx = find_attr_idx(die, DW_AT_type);
 		if (type_idx != -1) {
-			parse_type(ctx, die->attr_values[type_idx].reference, strbuf, size);
+			parse_type(ctx, die->attr_values[type_idx].reference, strbuf, size, visited);
 		}
 		rz_strbuf_append(strbuf, " &");
 		break;
 	default:
 		break;
 	}
+	set_u_delete(visited, offset);
```

Comment:
so is `visited` just to avoid infinite recursion? what happens if you don't remove it from `visited`?

Reply:
Good question, in this case it won't do any harm to not remove it because there is only at most one recursive call in every execution of this function.
The most "correct" way though to do it though is to delete it here so `visited` is always exactly the offsets on the stack. This becomes important if maybe later structs for example are recursed into here too since structs can have multiple members.
I could remove this delete call and add a comment instead. What do you think?

Reply:
Is it right to re-visit something that was already visited, even if not in the currently call stack?

Reply:
Yes. For example from the following struct:
```c
struct {
    A *member_a;
    A *member_b;
};
```
If you would not remove from the visited set every time you return from the function, as soon as you arrive at `member_b`, `A` would for example already be in the visited list from `member_a` and you would incorrectly detect an infinite recursion even though everything is fine.
But as said, such cases currently can't happen so it does not matter whether we delete or not.

Reply:
I think we should leave the `set_u_delete()` line in. It should become necessary as soon as we fully parse function types.